### PR TITLE
5145 item variant on all incoming stock

### LINKED ISSUE #5145: Capture item variant on all incoming stock
## Is your feature request related to a problem? Please describe 👀

- Add extra column with a dropdown input (before batch) anywhere stock comes in/changes:
  - Inbound Shipment
  - Customer Return
  - Stocktake
  - View stock detail
  - Inventory adjustment
  - Repack
- Add item_variant_id to stock_line, and probably invoice_line (mSupply change!) & sync

Screenshot 2024-10-16 at 2 12 36 PM

### Additional context 💌

At this point, we're just capturing the variant on the stock line to calculate current storage capacity used. i.e. we don't need to change anything with regards to distribution based on # packs per variant at this stage

### Moneyworks Jobcode 🧰

CCEIDEV

### PR DESCRIPTION
**Pull request by lache-melvin**
Fixes #5145

# 👩🏻‍💻 What does this PR do?

Capture item variant on incoming stock:

Screenshot 2024-10-30 at 1 11 19 PM

Also changes the `updates` to accept nullable input (to be able to unset the variant)

## 💌 Any notes for the reviewer?

Put the variant column fairly early in the input tables (after batch) per the designs, but do you think it should be more to the right/deprioritised column at this point?

Also, should I add a column to the view stock list view?

# 🧪 Testing

- [ ] Have some item variants configured for an item

For each of the below, try add and remove item variants:

- [ ] Create & edit customer return
- [ ] Create inbound shipment
- [ ] Create new stock line from View Stock
- [ ] Edit stock line from view stock

# 📃 Documentation

- [x] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

**Comment by jmbrunskill** [comment]
Looks like it might have been a pre-existing problem but seems bad that quantity to return is off the screen :(
image

File: client/packages/system/src/Item/api/operations.graphql [hunk 1]
```diff
@@ -266,6 +266,29 @@ query itemById($storeId: String!, $itemId: String!) {
   }
 }
 
+fragment ItemVariantOption on ItemVariantNode {
```
Comment (jmbrunskill) [inline client/packages/system/src/Item/api/operations.graphql:269]: Nice, feels like a potential pattern to re-use?

File: server/service/src/invoice_line/stock_in_line/insert/validate.rs [hunk 2]
```diff
@@ -31,6 +31,12 @@ pub fn validate(
         return Err(LocationDoesNotExist);
     }
 
+    if let Some(item_variant_id) = &input.item_variant_id {
+        if !check_item_variant_exists(connection, item_variant_id)? {
```
Comment (jmbrunskill) [inline server/service/src/invoice_line/stock_in_line/insert/validate.rs:35]: I guess to be thorough we should check that it not only exists but that it's for the right `item_id`?
Not overly worried though...

File: server/service/src/invoice_line/stock_in_line/update/validate.rs [hunk 3]
```diff
@@ -51,6 +52,17 @@ pub fn validate(
     if !check_location_exists(connection, store_id, &input.location)? {
         return Err(LocationDoesNotExist);
     }
+    // TODO
```
Comment (jmbrunskill) [inline server/service/src/invoice_line/stock_in_line/update/validate.rs:55]: Is this TODO correct?

File: server/service/src/invoice_line/stock_in_line/update/validate.rs [hunk 4]
```diff
@@ -51,6 +52,17 @@ pub fn validate(
...
+    match &input.item_variant_id {
+        Some(NullableUpdate {
+            value: Some(item_variant_id),
+        }) => {
+            if !check_item_variant_exists(connection, item_variant_id)? {
+                return Err(ItemVariantDoesNotExist);
+            }
+        }
+        _ => {}
+    }
```
Comment (jmbrunskill) [inline server/service/src/invoice_line/stock_in_line/update/validate.rs:65]:
```suggestion
    match &input.item_variant_id {
        Some(NullableUpdate {
            value: Some(item_variant_id),
        }) => {
            if !check_item_variant_exists(connection, item_variant_id)? {
                return Err(ItemVariantDoesNotExist);
            }
        }
        _ => {} //  We don't need to check item_variant if it's not being update, or if it's being updated to None
    }
```

I feel like this is sufficiently complicated to justify a comment? Took me a bit to parse it, possibly dad brain slowness though haha.

**Comment by lache-melvin** [comment]
> Looks like it might have been a pre-existing problem but seems bad that quantity to return is off the screen :( 

Yup :/ #3987

> Also related, and I guess highly debated, but it feels wrong that the customer return under distribution, is actually adding to your stock, vs other transactions under this heading which are reducing it.
> 
> I was confused as I was looking under replenishment...

And yep, this has gone around many a time on Telegram, grouping by "Incoming stock"/"Outgoing stock" vs "Customer related"/"Supplier related" 🤷‍♀️

**Review by jmbrunskill — APPROVED** [review]
Nice! Working well, merge away when ready!

One thought I had was that it might be nice to hide the item variant inputs if there's no variants configured. Basically so as to not change the UI  and save screen realestate for people who aren't using them, which might be a bulk of our users? A global pref might be nice?

Worthing discussing or adding to backlog @mark-prins ?

Personally I like the idea of `item_variant` showing up in the `ViewStock` list view, but also happy for that to come along in future if asked for :)

File: client/packages/system/src/Item/api/operations.graphql (hunk 1 above)
Comment (lache-melvin) [reply client/packages/system/src/Item/api/operations.graphql:269]: I wondered about it! In cases where we are doing a query to populate the options, this is nicer than having to map in the component - but I think many queries are reused from another list... arguably that's not great though!

File: server/service/src/invoice_line/stock_in_line/update/validate.rs (hunk 3 above)
Comment (lache-melvin) [reply server/service/src/invoice_line/stock_in_line/update/validate.rs:55]: Jaja it was at one point... 🙈

**Comment by lache-melvin** [comment]
> One thought I had was that it might be nice to hide the item variant inputs if there's no variants configured. Basically so as to not change the UI and save screen realestate for people who aren't using them, which might be a bulk of our users? A global pref might be nice?
> 
> Worthing discussing or adding to backlog @mark-prins ?

Yeah we had this for pack variants, I think it's nice to hide for majority of users, especially while they don't do much! It is more work though... let's see what Mark thinks 🙏 

> Personally I like the idea of `item_variant` showing up in the `ViewStock` list view, but also happy for that to come along in future if asked for :)

Was going to do if it was a 2 min job, but I think needs a new item_variant_by_id loader etc., think we can do later as needed 😁

**Pull request merged**